code/modules/vehicles/pimpin_ride.dm

```diff
@@ -1,76 +1,147 @@
 		update_appearance()
-	else if(mybag)
-		mybag.attackby(I, user)
+	else if(trash_bag)
+		trash_bag.attackby(I, user)
 	else
 		return ..()
 
 /obj/vehicle/ridden/janicart/update_overlays()
 	. = ..()
-	if(mybag)
+	if(trash_bag)
 		. += "cart_garbage"
-	if(floorbuffer)
-		. += "cart_buffer"
+	if(installed_upgrade)
+		var/mutable_appearance/overlay = new(SSgreyscale.GetColoredIconByType(installed_upgrade.overlay_greyscale_config, installed_upgrade.greyscale_colors))
+		overlay.icon_state = "janicart_upgrade"
+		. += overlay
 
 /obj/vehicle/ridden/janicart/attack_hand(mob/user, list/modifiers)
 	. = ..()
-	if(. || !mybag)
+	if(. || !trash_bag)
 		return
-	mybag.forceMove(get_turf(user))
-	user.put_in_hands(mybag)
-	mybag = null
+	trash_bag.forceMove(get_turf(user))
+	user.put_in_hands(trash_bag)
+	trash_bag = null
+	SEND_SIGNAL(src, COMSIG_VACUUM_BAG_DETACH)
 	update_appearance()
 
 /obj/vehicle/ridden/janicart/upgraded
-	floorbuffer = TRUE
+	installed_upgrade = new /obj/item/janicart_upgrade/buffer
+
+/obj/vehicle/ridden/janicart/upgraded/vacuum
+	installed_upgrade = new /obj/item/janicart_upgrade/vacuum
+
+/**
+ * # Janicart Upgrade
+ *
+ * Functional upgrades that can be installed into a janicart.
+ *
+ */
+/obj/item/janicart_upgrade
+	name = "base upgrade"
+	desc = "An abstract upgrade for mobile janicarts."
+	icon_state = "janicart_upgrade"
+	greyscale_config = /datum/greyscale_config/janicart_upgrade
+	/// The greyscale config for the on-cart installed upgrade overlay
+	var/overlay_greyscale_config = /datum/greyscale_config/janicart_upgrade/installed
+
+/**
+ * Called when upgrade is installed into a janicart
+ *
+ * Arguments:
+ * * installee - The cart the upgrade is being installed into
+ */
+/obj/item/janicart_upgrade/proc/install(obj/vehicle/ridden/janicart/installee)
+	return FALSE
+
+
+/**
+ * Called when upgrade is uninstalled from a janicart
+ *
+ * Arguments:
+ * * installee - The cart the upgrade is being removed from
+ */
+/obj/item/janicart_upgrade/proc/uninstall(obj/vehicle/ridden/janicart/installee)
+	return FALSE
+
+/obj/item/janicart_upgrade/buffer
+	name = "floor buffer upgrade"
+	desc = "An upgrade for mobile janicarts which adds a floor buffer functionality."
+	greyscale_colors = "#ffffff#6aa3ff#a2a2a2#d1d15f"
+
+/obj/item/janicart_upgrade/buffer/install(obj/vehicle/ridden/janicart/installee)
+	installee._AddElement(list(/datum/element/cleaning))
+
+/obj/item/janicart_upgrade/buffer/uninstall(obj/vehicle/ridden/janicart/installee)
+	installee._RemoveElement(list(/datum/element/cleaning))
+
+/obj/item/janicart_upgrade/vacuum
+	name = "vacuum upgrade"
+	desc = "An upgrade for mobile janicarts which adds a vacuum functionality."
+	greyscale_colors = "#ffffff#ffea6a#a2a2a2#d1d15f"
+
+/obj/item/janicart_upgrade/vacuum/install(obj/vehicle/ridden/janicart/installee)
+	installee._AddComponent(list(/datum/component/vacuum))
+	if (installee.trash_bag)
+		SEND_SIGNAL(installee, COMSIG_VACUUM_BAG_ATTACH, installee.trash_bag)
```

Comment:
This is kinda hacky if anything else would listen to the signal. I think it would be cleaner to just update it directly.